### DISCUSSION TITLE
796. Rotate String

### DISCUSSION THREAD

**Discussion by mah-shamim** [mah-shamim/leet-code-in-php#783] [Q&A] [answered]
**Topics:** `Mid Level`, `String`, `String Matching`, `Weekly Contest 75`

Given two strings `s` and `goal`, return _`true` if and only if `s` can become `goal` after some number of **shifts** on `s`_.

A **shift** on `s` consists of moving the leftmost character of `s` to the rightmost position.

- For example, if `s = "abcde"`, then it will be `"bcdea"` after one shift.

**Example 1:**

- **Input:** s = "abcde", goal = "cdeab"
- **Output:** true

**Example 2:**

- **Input:** s = "abcde", goal = "abced"
- **Output:** false

**Example 3:**

- **Input:** s = "a", goal = "a"
- **Output:** true

**Example 4:**

- **Input:** s = "ab", goal = "abc"
- **Output:** false

**Constraints:**

- `1 <= s.length, goal.length <= 100`
- `s` and `goal` consist of lowercase English letters.

**Comment by mah-shamim** [comment, marked as answer]
We can take advantage of the properties of string concatenation. Specifically, if we concatenate the string `s` with itself (i.e., `s + s`), all possible rotations of `s` will appear as substrings within that concatenated string. This allows us to simply check if `goal` is a substring of `s + s`.

### Approach

- Compare lengths of `s` and `goal` — if different, return `false` immediately.
- Concatenate `s` with itself to form `s + s`.
- Check if `goal` is a substring of `s + s` using `strpos`.
- Return `true` if found, `false` otherwise.

Let's implement this solution in PHP: **[796. Rotate String](https://github.com/mah-shamim/leet-code-in-php/tree/main/algorithms/000796-rotate-string/solution.php)**

```php
<?php
/**
 * @param String $s
 * @param String $goal
 * @return Boolean
 */
function rotateString(string $s, string $goal): bool
{
    // Check if lengths are equal; if not, return false
    if (strlen($s) !== strlen($goal)) {
        return false;
    }

    // Concatenate s with itself
    $doubleS = $s . $s;

    // Check if goal is a substring of doubleS
    return strpos($doubleS, $goal) !== false;
}

// Test cases
echo rotateString("abcde", "cdeab") . "\n";             // Output: true
echo rotateString("abcde", "abced") . "\n";             // Output: false
echo rotateString("a", "a") . "\n";                     // Output: true 
echo rotateString("ab", "abc") . "\n";                  // Output: false 
?>
```

### Explanation:

- **Length Check**: If lengths differ, rotation is impossible.
- **Double String**: Any rotation of `s` appears as a contiguous block in `s + s`.
- **Substring Check**: Searching for `goal` in `s + s` covers all possible rotations in O(n) time with built-in functions.
- **Edge Cases**: Handles empty strings (though constraints say length ≥ 1) and single-character strings correctly.

### Complexity:
- **Time Complexity**: _**O(n)**_ — `strpos` scans through `s + s` (length 2n) for `goal`.
- **Space Complexity**: _**O(n)**_ — for storing the concatenated string `s + s`.

**Reply by kovatz** [reply]
Thanks to solve the problem of determining if one string can become another through a series of shifts

**Reply by mah-shamim** [reply]
Thanks
